cloudevents/working-drafts/dds-format.md

```diff
@@ -0,0 +1,247 @@
+elements defined in the CloudEvents specification are are represented using
+a protobuf schema.
+
+The [Attributes](#2-attributes) section describes the naming conventions and
+data type mappings for CloudEvent attributes for use as protobuf message
+properties.
+
+The [Data](#3-data) section describes how the event payload is carried.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2 Content-Type
+
+There is no official IANA *media-type* designation for DDS, as such this
+specification uses 'application/dds' to identify such content.
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are represented in the DDS
+[schema][dds-schema].
+
+## 2.1 Type System
+
+The CloudEvents type system is mapped to IDL datatypes as follows :
+
+| CloudEvents   | DDS |
+| ------------- | ---------------------------------------------------------------------- |
+| Boolean       | [boolean][dds-schema] |
+| Integer       | [int32][dds-schema] |
+| String        | [string][dds-schema] |
+| Binary        | [bytes][dds-schema] |
+| URI           | [string][dds-schema] following [RFC 3986 §4.3][rfc3986-section43]|
+| URI-reference | [string][dds-schema] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [Timestamp][dds-schema]  |
+
+## 2.3 REQUIRED Attributes
+
+REQUIRED attributes are represented explicitly as fields in the [Event type definition][dds-schema].
+
+## 2.4 OPTIONAL Attributes & Extensions
+
+OPTIONAL and extension attributes are represented using a sequence of key-value
+constructs (DDS-CE-Attributes) enabling direct support of the CloudEvent
+[type system][ce-types]. An underlying union and enumeration data structure
+are needed to construct the key-value sequence to establish the association
+between the IDL types and the CloudEvent type system.
+
+```xml
+
+  <enum name="AttributeType">
+   <enumerator name="BOOL"/>
+   <enumerator name="INT32"/>
+   <enumerator name="STRING"/>
+   <enumerator name="BYTES"/>
+   <enumerator name="URI"/>
+   <enumerator name="URI_REF"/>
+   <enumerator name="TIMESTAMP"/>
+ </enum>
+	  
+ <union name="AttributeValue" extensibility="final">
+   <discriminator type="nonBasic" nonBasicTypeName="io::cloudevents::AttributeType"/>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::BOOL)"/>
+     <member name="ce_boolean" type="boolean"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::INT32)"/>
+     <member name="ce_integer" type="int32"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::STRING)"/>
+     <member name="ce_string" type="string" stringMaxLength="255"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::BYTES)"/>
+     <member name="ce_bytes" type="byte" sequenceMaxLength="100"/>
```

Comment:
nit: The CloudEvent type system calls this 'binary' not 'bytes'.

Additionally I don't believe there's any mention of max-length in the spec, rightly or wrongly. So I'm not sure what to suggest as a length constraint if that is required.

Reply:
We thought about it and at this time we're not ready to support optional and extension attributes yet, since there might be potential conflicts with what is defined at the DDS protocol level in QoS settings. So this section is going away.

Reply:
Responding to: Can I format an event using DDS, send it over HTTP, and de-serialize it back into a CE

The DDS format is tightly coupled to the DDS protocol. So when you send a message using the DDS protocol, you have to define the message format in DDS IDL, run the IDL through a code generator, and then get a set of classes that serialize/deserialize to the DDS wire protocol. 

I don't think there's really a way to use that serialization/deserialization code outside of DDS, the way you can with protobuf. So the end result is that the DDS format works only for the DDS protocol. I agree with you - this is an important point and I will add it to the doc.

Reply:
Apologies for the slow response...

If the formats & transports are tightly coupled I'm concerned it might not be appropriate to make a part of the CE specification.

If I'm not able to plug-and-play formats (wire-encoding) and different transport protocols this _may_ be better presented as a separate document that describes how CE's fit, or are used, in the context of DDS.

Ideally we should be able to realize a flow such as this (where Int == 'Intermediary'):

``` text
Source --[HTTP]--> Intermediary --[DDS]--> Intermediary --[AMQP]--> Intermediary --[KAFKA]--> SInk
```

Where the CE received by the Sink is equivalent to the CE emitted by the source, and the 'data'  is exactly what the source put into the original CE.

If we're not able to slot DDS into that type of pipeline I'm not sure this is the right place.

Apologies if I'm still not joining the (mental) dots.

Reply:
This is a valid concern and, honestly, I was having this same doubt after watching Doug's demo showing how CloudEvents is able to pass messages across protocols. If a message goes to DDS that it is not able to deserialize, it does not look at any other information in the message to determine who to forward to. It simply drops the message. Now, there could be updates made at the protocol level to cause this kind of forwarding behavior but it does not exist in the standard today. There is a DDS Web Integration standard that defines how DDS traffic can be tunneled over HTTP, but nothing that does the reverse type of action.

So this might be the blocking point. :-(